symengine/tests/eval/test_lambda_double.cpp

```diff
@@ -227,7 +232,18 @@ TEST_CASE("Check llvm and lambda are equal", "[llvm_double]")
     std::cout << std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1)
                      .count()
               << "us" << std::endl;
-    std::cout << d << " " << d2 << std::endl;
+
+    t1 = std::chrono::high_resolution_clock::now();
+    for (int i = 0; i < 500; i++) {
+        d3 = v3.call({0.4, 2.0, 3.0});
+    }
+    t2 = std::chrono::high_resolution_clock::now();
+    std::cout << std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1)
+                     .count()
+              << "us" << std::endl;
```

Comment:
This gives,
```
112052us   (Lambda)
35001us     (LLVM)
3620us       (LLVM with CSE)
```

I guess I was wrong. LLVM's CSE is not good and we can definitely do better. I wonder whether that's because we are assuming associativity in floating point arithmetics, while LLVM is not doing that.

Reply:
I think our CSE is a (lot) more expensive to do, and LLVM as well as gcc only do things that are quick to compile.

Reply:
As a rule of thumb, it's a good idea to do all these expression optimization before passing to gcc or LLVM.

Reply:
> I think our CSE is a (lot) more expensive to do, and LLVM as well as gcc only do things that are quick to compile.

I checked and `CSE + LLVM-no-CSE` is faster than `no-CSE + LLVM-with-CSE`.
`CSE+LLVM-with-CSE` is also faster than `no-CSE + LLVM-with-CSE` when the expressions are larger.

Reply:
Perfect. By "expensive" I meant expensive to compile, not to run. When you say faster, I think you only measure the actual evaluation, but not the construction, correct? It would be also interesting to compare how long it takes to construct (=compile) it.

Reply:
Compilers generally strike a medium between how fast it runs, and how slow it compiles. But we can spend an order of magnitude more on compiling it (i.e. more expensive CSE), in order to get faster runtime.

Reply:
My last comment was about construction. (My first comment was about evaluation).

Reply:
Can you put the timings into a table like this please:

| method | construction | evaluation |
| --------- | ------| ------------------------ |
| no-CSE + LLVM-no-CSE | | |
| CSE + LLVM-no-CSE | | |
| no-CSE + LLVM-with-CSE | | |
| CSE+LLVM-with-CSE | | |

I would be very interested in that. Possibly one table for a smaller expression and one table for a larger expression.

Reply:
This is for the expression `(log(x) + tan(x) + asin(x) + acos(x) + atan(x) + sinh(x) + cosh(x) + tanh(x) + asinh(y) + acosh(y) + atanh(x) + erf(x) + erfc(x) + gamma(x) + loggamma(x) + abs(x))*(2*y**4*z + sin(x)**2 + sin(x))*((log(x) + tan(x) + asin(x) + acos(x) + atan(x) + sinh(x) + cosh(x) + tanh(x) + asinh(y) + acosh(y) + atanh(x) + erf(x) + erfc(x) + gamma(x) + loggamma(x) + abs(x))*(2*y**4*z + sin(x)**2 + sin(x)) + 2**E + x**(exp(cos(x))))`

| method | construction (in us) | 500 evals (in us) |
| --------- | ------| ------------------------ |
| no-CSE + LLVM-no-CSE | 41,263 | 2,375 |
| CSE + LLVM-no-CSE |20,643 | 1,269 |
| no-CSE + LLVM-with-CSE |46,310 | 2,316|
| CSE+LLVM-with-CSE | 20,990| 1,267 |

Reply:
So on this expression, LLVM-with-CSE does little over LLVM-no-CSE in terms of runtime, but it slows things down when constructing. Our CSE makes things 2x faster to construct, and gives almost 2x speedup at runtime.

Reply:
Yes. (Btw, difference between `2,375` and `2,316` is negligible. I mean, some runs have the timings flipped for those 2. I just gave the runtimes for one. Didn't have time to run this multiple times and give averages.)

Reply:
With this in mind, should we enable `cse` by default?

Reply:
I think so, I would enable it by default. Later, we should test this for more expressions.

Reply:
Well, we should also run a quick benchmark on an expression that can't benefit from CSE, to see how we do there. Because perhaps there is an overhead there. The expression above has large benefits from CSE.

Reply:
Yes, when there's no benefit from CSE, there's a small overhead.